### PR TITLE
Extend `RUF005` to recursive and literal-literal concatenations

### PR DESCRIPTION
**Pull request by hoel-bagard**
This attempts to fix https://github.com/charliermarsh/ruff/issues/3496.

**It introduces the following changes:**
- The building of the new suggested expression is moved to a recursive function to handle cases with multiple concatenations on a single line.
- The rule now also checks for concatenations of lists/tuples (for example `a = [1] + [2]`).  (I'm not sure if this behavior is desired, it's not directly related to the issue).
- New tests to check the new behavior.

### CODE REVIEW

File: crates/ruff/src/rules/ruff/rules/collection_literal_concatenation.rs [hunk 1]
```diff
@@ -54,80 +54,135 @@ enum Kind {
     Tuple,
 }
 
-/// RUF005
-/// This suggestion could be unsafe if the non-literal expression in the
-/// expression has overridden the `__add__` (or `__radd__`) magic methods.
-pub(crate) fn collection_literal_concatenation(checker: &mut Checker, expr: &Expr) {
+fn build_new_expr(expr: &Expr) -> Option<Expr> {
```
Comment (konstin) [inline crates/ruff/src/rules/ruff/rules/collection_literal_concatenation.rs:57]: Please add a comment what this function does

File: crates/ruff/src/rules/ruff/rules/collection_literal_concatenation.rs [hunk 2]
```diff
@@ -54,80 +54,135 @@ enum Kind {
...
+    else if splat_element.is_list_expr() || splat_element.is_tuple_expr() {
+        new_elts = other_elements.clone();
+
+        let mut elts = match splat_element {
+            Expr::List(ast::ExprList { elts, .. }) => elts.clone(),
+            Expr::Tuple(ast::ExprTuple { elts, .. }) => elts.clone(),
+            _ => return None,
+        };
+
+        if splat_at_left {
```
Comment (konstin) [inline crates/ruff/src/rules/ruff/rules/collection_literal_concatenation.rs:134]: In what case would `splat_at_left` be true here?

File: crates/ruff/src/rules/ruff/rules/collection_literal_concatenation.rs [hunk 3]
```diff
@@ -54,80 +54,135 @@ enum Kind {
...
+        || splat_element.is_attribute_expr()
     {
-        return;
+        new_elts = make_splat_elts(splat_element, other_elements, splat_at_left);
+    }
+    // If the splat element is itself a list/tuple, insert them in the other list/tuple.
+    else if splat_element.is_list_expr() || splat_element.is_tuple_expr() {
+        new_elts = other_elements.clone();
+
+        let mut elts = match splat_element {
```
Comment (konstin) [inline crates/ruff/src/rules/ruff/rules/collection_literal_concatenation.rs:128]: How would this deal with `(1,) + [2]` or `[1] + (2,)`?

Comment (hoel-bagard) [reply crates/ruff/src/rules/ruff/rules/collection_literal_concatenation.rs:128]: I did think about this, but 

Running ruff on a file with:
```python
a = (1,) + [2]
b = (1, 2, 3) + (4,)
c = (7, 8, 9) + b
```

The current version outputs a warning only for the last line, whereas the PR in its current state outputs a warning for all three lines. While a warning on the first line is clearly undesired (that's a typing error), should the one on the second line be kept ?

File: crates/ruff/src/rules/ruff/rules/collection_literal_concatenation.rs (hunk 2 above)
Comment (hoel-bagard) [reply crates/ruff/src/rules/ruff/rules/collection_literal_concatenation.rs]: It seems like it's always true indeed, I've removed the `if`, thank you!

**Comment by charliermarsh** [comment]
Do you mind adding some information to the PR summary around what needed to change, and why?

**Comment by hoel-bagard** [comment]
To be honest I'm not sure I went about it the right way, so if you think that there is a better way to solve the issue I don't mind redoing it another way or trying to tackle a different issue.

On another note, I have a question about the formatting of the test fixture. The following test always gets formatted by black during the pre-commit verify:
```python
foo + [  # This will be preserved, but doesn't prevent the fix
]
```

Should I remove this test (and maybe include a replacement one), or if it better to use `--no-verify` when making a commit modifying the test fixture file ?

File: crates/ruff/src/rules/ruff/rules/collection_literal_concatenation.rs [hunk 4]
```diff
@@ -48,86 +48,128 @@ fn make_splat_elts(
     new_elts
 }
 
-#[derive(Debug, Copy, Clone)]
+#[derive(Debug, Copy, Clone, PartialEq)]
```
Comment (MichaReiser) [inline crates/ruff/src/rules/ruff/rules/collection_literal_concatenation.rs:51]:
```suggestion
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
```

Adding `Eq` here should be fine.

Comment (hoel-bagard) [reply crates/ruff/src/rules/ruff/rules/collection_literal_concatenation.rs:51]: @MichaReiser Thank you for the comment.

Adding `Eq` does not cause any issue indeed. I'm a bit new to Rust, is it good practice to add `Eq` along with `PartialEq` wherever possible ?

**Review by konstin — APPROVED** [review]
This needs to be updated with the changes from main and a `cargo test` `cargo insta review`, otherwise i think it's ready

**Comment by charliermarsh** [comment]
I'll take a look at this tonight and (hopefully) merge.

**Pull request merged**